### DISCUSSION TITLE
Error when an endpoint tries to return a generator / map object

### DISCUSSION THREAD

**Discussion by OnePunchMango** [fastapi/fastapi#6447] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
@app.get("/records", response_model=Iterable[RecordModel])
async def get_records() -> Iterable[RecordModel]:
    # {queried_records} is a generator returned from the database query
    queried_records = get_records_from_database() 
    formatted_records = map(lambda record: __format(record), queried_records)
    return formatted_records

def __format(queried_record: Dict[str, Union[str, HttpUrl]) -> Union[RecordModel, None]:
    formatted_record = RecordModel(
        key_1 = queried_record[key_a],
        key_2 = queried_record[key_b],
        key_3 = queried_record[key_c]
    )
    return formatted_record
```

### Description

Hey! I'm working on my first FastAPI and Pydantic project, came across a problem when trying to return a generator from an endpoint. The problem is below, any inputs will be really appreciated!

I have an API endpoint where I first get data records from a database, then format each record using a map function. However, when returning the mapped result to the endpoint, the key-value mapping went wrong. Note, I'd like to keep the return type from the endpoint as a generator for performance sake (big data volume).

From the pseudo-code in this post, I got an error when running the endpoint

`ValueError: [ValueError('dictionary update sequence element #0 has length 3; 2 is required'), TypeError('vars() argument must have __dict__ attribute')]`

What do you think about the ValueError mentioned above? Big thanks for your inputs in advance!

### Operating System

macOS

### FastAPI Version

0.68.0

### Python Version

3.9

**Comment by Dustyposa** [comment, marked as answer]
this should work:
```python
@app.get("/records", response_model=Iterable[RecordModel])
async def get_records() -> Iterable[RecordModel]:
    # {queried_records} is a generator returned from the database query
    queried_records = get_records_from_database() 
    formatted_records = map(lambda record: __format(record), queried_records)
    return list(formatted_records)

def __format(queried_record: Dict[str, Union[str, HttpUrl]) -> Union[RecordModel, None]:
    formatted_record = RecordModel(
        key_1 = queried_record[key_a],
        key_2 = queried_record[key_b],
        key_3 = queried_record[key_c]
    )
    return formatted_record
```
In http protocol you need directly  return response body.
If you want to know how to send large data in rest api, you can limit the response data size, eg: [link](https://stackoverflow.com/questions/11725078/restful-api-handling-large-amounts-of-data)

**Comment by OnePunchMango** [comment]
Thanks @Dustyposa! I took your suggestion and it works for my use case now 😃

**Comment by ricardo-reis-1970** [comment]
I had the same issue today, but not with generators in general. I have a function that yields values, so it's a generator. This works very well. However, the returned values were big dictionaries and all I needed was the `'name'` key.

I used `map` and then it started giving the exact same error @yu-mao first described. I'm not sure why this happens inside FastAPI, but `map` objects seem to present a slightly different interface when compared to other generators.

I have two remarks regarding the original question:

## 1. Performance
Regarding the line:
```python
    queried_records = get_records_from_database() 
```
is this variable itself a generator? Because otherwise it will be accumulating a "big data volume".

## 2. Performance... again (aka. Tuple x List)
The solution proposed by @Dustyposa consisted of wrapping the result in a `list(...)`:
```python
    return list(formatted_records)
```

The exact same effect (API response) can be achieved with a tuple instead:
```python
    return tuple(formatted_records)
```
This is much lighter than `list`. In fact, throughout our code, all of us abuse `list` for cases where `tuple` would work as well, and overall much cheaper.

## 3. (yet another) Solution
In any of the previous cases, it felt like I was defeating the purpose of generators by imposing a rendering "stop-and-go2 stage upon the pipe. Be it `list` or `tuple`, the effect is the same: block the pipe until the data is all gathered and then deliver it as a list or a tuple.

So, instead, I thought of something I believe is both more elegant and more efficient:
```python
def yield_from(gen):
   yield from gen
```

This function will pass through any generator-like object (such as `map`) and morph it into a full-blown generator. Now, instead of doing:
```python
    return list(formatted_records)
```
we can do:
```python
    return yield_from(formatted_records)
```

This is equivalent to doing:
```python
    return (x for x in formatted_records)
```
except that in this last case, we use generator comprehension instead of a `yield from`.

I hope this helps.
